### PR TITLE
[Tune] Fix storage client creation when sync function tpl is not provided (#26714)

### PR DESCRIPTION
**Pull request by rohit-annigeri**
## Why are these changes needed?

It solves issue of restore from object failing to work during ray tune with sync config and PBT scheduler.

## Related issue number

Closes https://github.com/ray-project/ray/issues/26714

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Review by krfricke — CHANGES_REQUESTED** [review]
Thanks @rohit-annigeri - actually that part is correct. Your change will use the legacy sync client which is deprecated and even removed on master.

The correct fix here is to not try to sync the directory if it already exists locally. This regression actually exists in master as well. Here is a draft PR fixing it: https://github.com/ray-project/ray/pull/26725
I'll add tests to that PR to make sure this doesn't come up again and that everything works. 

Could you apply that patch to 1.13.0 and see if it fixes your problem? 

Thanks!

**Comment by rohit-annigeri** [comment]
```Could you apply that patch to 1.13.0 and see if it fixes your problem?```
So that would merge to releases/1.13.0 ?
Also verified that it does fix my problem.

**Comment by krfricke** [comment]
I mean if you apply the changes from #26725 to `releases/1.13.0`, does that also solve the problem?

The current fix in this PR disables syncing via pyarrow fs completely (as `self.uses_cloud_checkpointing` will be True), which is a regression.

The correct location to add the check would be at https://github.com/ray-project/ray/blob/releases/1.13.0/python/ray/tune/trainable.py#L536 to not trigger syncing if the directory already exists.

I can also update this PR if you prefer.

We will have to open it for `releases/1.13.1` in any case.

Thanks!

**Comment by rohit-annigeri** [comment]
I think I follow. I've updated the PR to reflect the change. Let me know if it makes sense. Not really sure if this is the best way to handle the exception

**Comment by rohit-annigeri** [comment]
One side affect while "fixing" this, now it seems to sync all tmp43save_to_object folders to s3 as well. 

image

File: python/ray/tune/trainable.py [hunk 1]
```diff
@@ -544,6 +546,11 @@ def restore(self, checkpoint_path: str, checkpoint_node_ip: Optional[str] = None
                 # Only keep for backwards compatibility
                 self.storage_client.sync_down(external_uri, local_dir)
                 self.storage_client.wait_or_retry()
+            elif os.path.exists(checkpoint_path):
+                try:
+                    TrainableUtil.find_checkpoint_dir(checkpoint_path)
+                except Exception:
+                    pass
```
Comment (krfricke) [inline python/ray/tune/trainable.py:553]: I think we can move this check earlier. Basically

```
checkpoint_exists = False

if os.path.exists(checkpoint_path):
    try:
        TrainableUtil.find_checkpoint_dir(checkpoint_path) 
    except Exception:
        pass
    else:
        checkpoint_exists = True

if checkpoint_exists:
    pass
elif self.uses_cloud_checkpointing:
    # ...
```

then we'll never download if the checkpoint already exists.

Comment (rohit-annigeri) [reply python/ray/tune/trainable.py]: agreed, made the change

**Review by krfricke — APPROVED** [review]
This looks good to me. We will merge this once we commence work on 1.13.1. I may update this PR with test cases when updating #26725

**Comment by rohit-annigeri** [comment]
hey @krfricke is this PR still good to go ?

**Comment by krfricke** [comment]
Generally yes, but there is currently no timeline for 1.13.1

**Pull request closed without merging**